src/toxcore/io_tokio.rs

```diff
@@ -1,17 +1,91 @@
 //! Defines `IoFuture` and `IoStream`
 
 use std::fmt::Debug;
-use std::io::{Error, ErrorKind};
+use std::io::{Error as IoError, ErrorKind as IoErrorKind};
 use std::time::Duration;
 
+use std::fmt;
+
+use failure::{Backtrace, Context, Fail};
 use futures::{Future, Sink, Stream};
 use tokio::util::FutureExt;
 
 /// A convenience typedef around a `Future` whose error component is `io::Error`
-pub type IoFuture<T> = Box<Future<Item = T, Error = Error> + Send>;
+pub type IoFuture<T> = Box<Future<Item = T, Error = IoError> + Send>;
 
 /// A convenience typedef around a `Stream` whose error component is `io::Error`
-pub type IoStream<T> = Box<Stream<Item = T, Error = Error> + Send>;
+pub type IoStream<T> = Box<Stream<Item = T, Error = IoError> + Send>;
+
+/// Error that can happen during sending all packet to peer.
+#[derive(Debug)]
+pub struct SendAllToError {
+    ctx: Context<SendAllToErrorKind>,
+}
+
+impl SendAllToError {
```

Comment:
Shouldn't it be named just `SendToError`? It seems `send_to` function will have the same error.
Also `send_to` function can't return `TimedOut` error. Maybe split error types for `send_to` and `send_to_bounded`?

Reply:
I think `TimedOut` error occurs during sending packet(s).
So, `SendToError` can include `TimedOut`.

Reply:
No, it can occur only if we set timeout ourselves.